sqlite/comdb2build.c

```diff
@@ -846,200 +873,260 @@ void comdb2RebuildIndex(Parse* pParse, Token* nm, Token* lnm, Token* index, int
-    
-    
     if (str) {
-        TokenStr(version, ver);
-        strcpy(sc->fname, version);
+        if (comdb2TokenToStr(ver, sp_version, sizeof(sp_version))) {
+            setError(pParse, SQLITE_MISUSE, "Procedure version is too long");
+            goto cleanup;
+        }
+        strcpy(sc->fname, sp_version);
     } else {
         sc->newcsc2 = malloc(ver->n + 1);
         strncpy(sc->newcsc2, ver->z, ver->n);
         sc->newcsc2[ver->n] = '\0';
     }
     sc->delsp = 1;
-  
-    comdb2prepareNoRows(v, pParse, 0, sc, &comdb2SqlSchemaChange_tran, 
-                        (vdbeFuncArgFree)  &free_schema_change_type);
+
+    comdb2prepareNoRows(v, pParse, 0, sc, &comdb2SqlSchemaChange_tran,
+                        (vdbeFuncArgFree)&free_schema_change_type);
+    return;
+
+cleanup:
+    free_schema_change_type(sc);
+    return;
 }
-/********************* PARTITIONS  **********************************************/
 
+/********************* PARTITIONS  **********************************************/
 
-void comdb2CreateTimePartition(Parse* pParse, Token* table, Token* partition_name, Token* period, Token* retention, Token* start)
+void comdb2CreateTimePartition(Parse* pParse, Token* table,
+                               Token* partition_name, Token* period,
+                               Token* retention, Token* start)
 {
     Vdbe *v  = sqlite3GetVdbe(pParse);
 
-    int max_length;
-
     BpfuncArg *arg = (BpfuncArg*) malloc(sizeof(BpfuncArg));
-    if (!arg) goto err; 
+    if (!arg) {
+        setError(pParse, SQLITE_NOMEM, "Out of Memory");
+        goto clean_arg;
+    }
     bpfunc_arg__init(arg);
 
     BpfuncCreateTimepart *tp = malloc(sizeof(BpfuncCreateTimepart));
-    if (!tp) goto err;
+    if (!tp) {
+        setError(pParse, SQLITE_NOMEM, "Out of Memory");
+        goto clean_arg;
+    }
     bpfunc_create_timepart__init(tp);
     
     arg->crt_tp = tp;
     arg->type = BPFUNC_CREATE_TIMEPART;
     tp->tablename = (char*) malloc(MAXTABLELEN);
-    memset(tp->tablename, '\0', MAXTABLELEN);
-    if (table && chkAndCopyTableTokens(v, pParse, tp->tablename, table, NULL, 1)) 
+    if (!tp->tablename) {
+        setError(pParse, SQLITE_NOMEM, "Out of Memory");
+        goto clean_arg;
+    }
+    if (table && chkAndCopyTableTokens(v, pParse, tp->tablename, table, NULL, 1))
         goto err;
 
-
-    max_length = partition_name->n < MAXTABLELEN ? partition_name->n : MAXTABLELEN;
     tp->partition_name = (char*) malloc(MAXTABLELEN);
-    memset(tp->partition_name, '\0', MAXTABLELEN);
-    strncpy(tp->partition_name, partition_name->z, max_length);
+    if (!tp->partition_name) {
+        setError(pParse, SQLITE_NOMEM, "Out of Memory");
+        goto clean_arg;
+    }
+    if (partition_name->n >= MAXTABLELEN) {
+        setError(pParse, SQLITE_MISUSE, "Partition name is too long");
+        goto clean_arg;
+    }
+    strncpy0(tp->partition_name, partition_name->z, partition_name->n + 1);
```

Comment:
maybe out of the scope of this PR, but I am wondering if we need dequote for partition name

Reply:
time partition and friends seem to dequote only table name, i think it is good to leave it as it is for now.